internal/scheduler/reports.go

```diff
@@ -271,41 +271,129 @@ func extractQueueAndJobContexts(sctx *schedulercontext.SchedulingContext) (map[s
 	return queueSchedulingContextByQueue, jobSchedulingContextByJobId
 }
 
+func (repo *SchedulingContextRepository) getSchedulingReportForQueue(queueName string) schedulingReport {
+	mostRecent, _ := repo.GetMostRecentQueueSchedulingContextByExecutor(queueName)
+	mostRecentSuccessful, _ := repo.GetMostRecentSuccessfulQueueSchedulingContextByExecutor(queueName)
+	mostRecentPreempting, _ := repo.GetMostRecentPreemptingQueueSchedulingContextByExecutor(queueName)
+
+	return schedulingReport{
+		mostRecentSchedulingContextByExecutor:           armadamaps.MapValues(mostRecent, schedulercontext.GetSchedulingContextFromQueueSchedulingContext),
+		mostRecentSuccessfulSchedulingContextByExecutor: armadamaps.MapValues(mostRecentSuccessful, schedulercontext.GetSchedulingContextFromQueueSchedulingContext),
+		mostRecentPreemptingSchedulingContextByExecutor: armadamaps.MapValues(mostRecentPreempting, schedulercontext.GetSchedulingContextFromQueueSchedulingContext),
+
+		sortedExecutorIds: repo.GetSortedExecutorIds(),
+	}
+}
+
+func (repo *SchedulingContextRepository) getSchedulingReportForJob(jobId string) schedulingReport {
```

Comment:
The existing data structures are a little awkward in this case. We keep around scheduling contexts, queue scheduling contexts, and job scheduling contexts, all of which cover a different subset of jobs; job scheduling contexts don't cover preempted jobs. I imagine that we'd want to fall back on scheduling and queue scheduling contexts even if we made `mostRecentJobSchedulingContextByExecutorByJobId` cover preempted jobs.

Reply:
Job scheduling contexts should cover preempted jobs. In the sense that they cover the most recent attempt to re-schedule a job, which would've failed for a preempted job. I think it'll be cleaner if we just store the scheduling contexts indexed by queue and jobId. Then, whenever we want a queue or job report, we just look inside the relevant scheduling context.